ubenchmark/src/jmh/java/org/apache/calcite/benchmarks/MetadataBenchmark.java

```diff
@@ -0,0 +1,140 @@
+            + " \"sales_fact_1997\" as \"sales_fact_1997\",\n"
+            + " \"time_by_day\" as \"time_by_day\",\n"
+            + " \"product_class\" as \"product_class\",\n"
+            + " \"product\" as \"product\"\n"
+            + "where \"sales_fact_1997\".\"store_id\" = \"store\".\"store_id\"\n"
+            + "and \"store\".\"store_country\" = 'USA'\n"
+            + "and \"sales_fact_1997\".\"time_id\" = \"time_by_day\".\"time_id\"\n"
+            + "and \"time_by_day\".\"the_year\" = 1997\n"
+            + "and \"sales_fact_1997\".\"product_id\" = \"product\".\"product_id\"\n"
+            + "and \"product\".\"product_class_id\" = \"product_class\".\"product_class_id\"\n"
+            + "group by \"store\".\"store_country\",\n"
+            + " \"time_by_day\".\"the_year\",\n"
+            + " \"product_class\".\"product_family\"")
+        .withHook(Hook.CONVERTED, (Consumer<RelNode>) rel -> {
+          rel.getCluster().setMetadataQuerySupplier(supplier);
+          rel.getCluster().invalidateMetadataQuery();
+        })
+        .explainContains(""
+            + "EnumerableAggregate(group=[{1, 6, 10}], m0=[COUNT()])\n"
+            + "  EnumerableMergeJoin(condition=[=($2, $8)], joinType=[inner])\n"
+            + "    EnumerableSort(sort0=[$2], dir0=[ASC])\n"
+            + "      EnumerableMergeJoin(condition=[=($3, $5)], joinType=[inner])\n"
+            + "        EnumerableSort(sort0=[$3], dir0=[ASC])\n"
+            + "          EnumerableHashJoin(condition=[=($0, $4)], joinType=[inner])\n"
+            + "            EnumerableCalc(expr#0..23=[{inputs}], expr#24=['USA':VARCHAR(30)], "
+            + "expr#25=[=($t9, $t24)], store_id=[$t0], store_country=[$t9], $condition=[$t25])\n"
+            + "              EnumerableTableScan(table=[[foodmart2, store]])\n"
+            + "            EnumerableCalc(expr#0..7=[{inputs}], proj#0..1=[{exprs}], "
+            + "store_id=[$t4])\n"
+            + "              EnumerableTableScan(table=[[foodmart2, sales_fact_1997]])\n"
+            + "        EnumerableCalc(expr#0..9=[{inputs}], expr#10=[CAST($t4):INTEGER], "
+            + "expr#11=[1997], expr#12=[=($t10, $t11)], time_id=[$t0], the_year=[$t4], "
+            + "$condition=[$t12])\n"
+            + "          EnumerableTableScan(table=[[foodmart2, time_by_day]])\n"
+            + "    EnumerableHashJoin(condition=[=($0, $2)], joinType=[inner])\n"
+            + "      EnumerableCalc(expr#0..14=[{inputs}], proj#0..1=[{exprs}])\n"
+            + "        EnumerableTableScan(table=[[foodmart2, product]])\n"
+            + "      EnumerableCalc(expr#0..4=[{inputs}], product_class_id=[$t0], "
+            + "product_family=[$t4])\n"
+            + "        EnumerableTableScan(table=[[foodmart2, product_class]])")
+        .returns("c0=USA; c1=1997; c2=Non-Consumable; m0=16414\n"
+            + "c0=USA; c1=1997; c2=Drink; m0=7978\n"
+            + "c0=USA; c1=1997; c2=Food; m0=62445\n");
```

Comment:
I agree in theory. In reality, I'm looking for baby steps here. 

Based on my reviewing of metadata tickets back to the original refactor Julian did several years ago, it appears that multiple developers of the metadata module would manually use this query to test the performance. I wanted to at least get it in a place where it could be run in a reliable/standard/consistent way. Realistically, we should move to something that is actually a microbenchmark. Having the entirety of query planning rule firing/transformations/etc being part of a test to understand metadata performance leads to a lot of noise in the measurement (beyond just the plan validation). 

For now, I thought it was better to keep this in since it also serves to confirm that the providers are behaving consistently.

Reply:
I am not saying to turn this into a real microbenchmark since we are quite far from there anyways. However, if we need to ensure correctness then wouldn't be better to add unit tests for this (if there are aren't already). Anyways minor comment feel free to ignore.